benchalerts/clients.py

```diff
@@ -0,0 +1,209 @@
+# Copyright (c) 2022, Voltron Data.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import abc
+import os
+from typing import Optional
+
+import requests
+from requests.adapters import HTTPAdapter
+from urllib3.util.retry import Retry
+
+from .log import fatal_and_log, log
+
+
+class _BaseClient(abc.ABC):
+    """A client to interact with an API.
+
+    Parameters
+    ----------
+    adapter
+        A requests adapter to mount to the requests session. If not given, one will be
+        created with a backoff retry strategy.
+    """
+
+    base_url: str
+    timeout_s = 10
+
+    def __init__(self, adapter: Optional[HTTPAdapter]):
+        if not adapter:
+            retry_strategy = Retry(
+                total=5,
+                status_forcelist=frozenset((429, 502, 503, 504)),
+                backoff_factor=4,  # will retry in 2, 4, 8, 16, 32 seconds
+            )
+            adapter = HTTPAdapter(max_retries=retry_strategy)
+
+        self.session = requests.Session()
+        self.session.mount("https://", adapter)
+
+    def get(self, path: str) -> dict:
+        url = self.base_url + path
```

Comment:
This is fine, but it'd be nice if it wasn't sensitive to whether `base_url` has a trailing `/` or not. Is there an equivalent to `os.path.join()` for URLs in requests or urllib3 or somewhere? Low-priority, just a nice-to-have

Reply:
Eh I guess this is very low priority bc `base_url` will always be set in child classes

Reply:
Luckily conbench doesn't care if you hit `https://conbench.biz/api/X` or `https://conbench.biz//api/X` so a trailing slash shouldn't matter. I just added an integration test to make sure of that. 😄 

If someone does `GithubRepoClient(repo="user/repo/")` then there will be a problem, but I doubt a script-implementer will do that and not understand the error!

Reply:
Did a quick search and found [urllib.parse.urljoin()](https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urljoin)

Reply:
That's a little tricky because it strips off the part of the path I was intentionally saving with the base URL:
```
>>> urljoin('https://conbench.biz/api', '/commits/')
'https://conbench.biz/commits/'
```

I think the `+` operator will be fine for now. If we add another client that can't use this pattern without risk of user error, we can revisit.

Reply:
That's a little tricky because of this behavior:
```
>>> urljoin('https://conbench.biz/api', '/commits/')
'https://conbench.biz/commits/'
>>> urljoin('https://conbench.biz/api/', '/commits/')
'https://conbench.biz/commits/'
>>> urljoin('https://conbench.biz/api/', 'commits/')
'https://conbench.biz/api/commits/'
>>> urljoin('https://conbench.biz/api', 'commits/')
'https://conbench.biz/commits/'
```
So we'd have to ensure the base_url always ends with a slash and the path never begins with a slash, in order to get everything to work the way we want. At that point it's not really much easier than the `+` operator.